pkpdapp/pkpdapp/models/inference.py

```diff
@@ -0,0 +1,356 @@
+#
+
+from django.db import models
+from django.db.models import Q
+from pkpdapp.models import (
+    Variable, Dataset, PharmacodynamicModel,
+    PharmacokineticModel, DosedPharmacokineticModel,
+    BiomarkerType)
+
+
+class Inference(models.Model):
+    """
+    An inference process.
+    """
+    description = models.TextField(
+        help_text='short description of what this inference does',
+        blank=True, default=''
+    )
+
+    datetime = models.DateTimeField(
+        help_text=(
+            'date/time the experiment was conducted. '
+            'All time measurements are relative to this date/time, ' +
+            'which is in YYYY-MM-DD HH:MM:SS format. For example, ' +
+            '2020-07-18 14:30:59'
+        ),
+        null=True, blank=True
+    )
+
+    class InferenceType(models.TextChoices):
+        SAMPLING = 'SA', 'Sampling'
+        OPTIMISATION = 'OP', 'Optimisation'
+
+    inference_type = models.CharField(
+        max_length=2,
+        choices=InferenceType.choices,
+        default=InferenceType.OPTIMISATION,
+    )
+
+    class SamplingAlgorithm(models.TextChoices):
+        HB = 'HB', 'Haario-Bardenet'
+        DE = 'DE', 'Differential evolution'
+        DR = 'DR', 'DREAM'
+        PO = 'PO', 'Population MCMC'
+
+    sampling_algorithm = models.CharField(
+        choices=SamplingAlgorithm.choices,
+        default=SamplingAlgorithm.HB,
+        help_text='sampling algorithm to use for inference')
+
+    class OptimisationAlgorithm(models.TextChoices):
+        CMAES = 'CMAES', 'CMAES'
+        XNES = 'XNES', 'XNES'
+        SNES = 'SNES', 'SNES'
+        PSO = 'PSO', 'PSO'
+        NM = 'NM', 'Nelder-Mead'
+
+    optimisation_algorithm = models.CharField(
+        choices=OptimisationAlgorithm.choices,
+        default=OptimisationAlgorithm.CMAES,
+        help_text='optimisation algorithm to use for inference')
+
+    number_of_iterations = models.IntegerField(
+        help_text='number of iterations'
+    )
+
+    time_elapsed = models.IntegerField(
+        default=0,
+        help_text='Elapsed run time for inference in seconds'
+    )
+
+    # potentially for optimisation too (as in number of starting points)
+    number_of_chains = models.IntegerField(
+        help_text='number of chains'
+    )
+
+    number_of_function_evals = models.IntegerField(
+        help_text='number of function evaluations'
+    )
+
+
+class StoredPharmacodynamicModel(PharmacodynamicModel):
+    """
+    Stored PD model.
+    """
+
+
+class StoredPharmacokineticModel(PharmacokineticModel):
+    """
+    Stored PK model.
+    """
+
+
+class DosedPharmacokineticModel(DosedPharmacokineticModel):
+    """
+    Stored dosed PK model.
+    """
+
+
+class ObjectiveFunction(Variable):
+    """
+    Abstract model class for objective functions.
+    """
+    biomarker_type = models.OnetoOne(
+        BiomarkerType,
+        on_delete=models.CASCADE,
+        blank=True, null=False)
+    pd_model = models.ForeignKey(
```

Comment:
unfortunately you can't override fields on a inherited model (https://stackoverflow.com/questions/2344751/in-django-model-inheritance-does-it-allow-you-to-override-a-parent-models-a/18159571#18159571)

Reply:
however, you can override fields on an abstract base